blockchain/checkpoints.go

```diff
@@ -27,31 +27,33 @@ func newHashFromStr(hexStr string) *chainhash.Hash {
 	return hash
 }
 
-// DisableCheckpoints provides a mechanism to disable validation against
-// checkpoints which you DO NOT want to do in production.  It is provided only
-// for debug purposes.
+// RemoveCheckpoints removes all checkpoints defined for this chain.
+// Use of this method is solely intended for debugging!
 //
 // This function is safe for concurrent access.
-func (b *BlockChain) DisableCheckpoints(disable bool) {
+func (b *BlockChain) RemoveCheckpoints() {
 	b.chainLock.Lock()
-	b.noCheckpoints = disable
+	b.checkpoints = make([]chaincfg.Checkpoint, 0)
+	b.checkpointsByHeight = make(map[int32]*chaincfg.Checkpoint)
 	b.chainLock.Unlock()
 }
 
-// Checkpoints returns a slice of checkpoints (regardless of whether they are
-// already known).  When checkpoints are disabled or there are no checkpoints
-// for the active network, it will return nil.
+// Checkpoints returns a slice of checkpoints.
+// When there are no checkpoints for the chain, it will return nil.
 //
 // This function is safe for concurrent access.
 func (b *BlockChain) Checkpoints() []chaincfg.Checkpoint {
 	b.chainLock.RLock()
-	defer b.chainLock.RUnlock()
-
-	if b.noCheckpoints || len(b.chainParams.Checkpoints) == 0 {
-		return nil
-	}
+	checkpoints := b.checkpoints
+	b.chainLock.RUnlock()
```

Comment:
Yes, the caller otherwise has no way of knowing the concurrency safety characteristics (or lack thereof) of the function.  How the function goes about achieving that, whether it be via a mutex, a channel, CAS, or use of immutable data structures doesn't change that.